src/str.rs

```diff
@@ -34,62 +35,46 @@ fn find_char_midpoint(chars: &str) -> usize {
 
 
 /// Parallel extensions for strings.
-///
-/// Implementing this trait is not permitted outside of `rayon`.
-pub trait ParallelString {
-    private_decl!{}
-
+pub trait ParallelString: Borrow<str> {
```

Comment:
Hmm, so I wonder -- is `Borrow` a better choice than `Deref`? The overlap between those two traits is fairly extreme, iirc the major difference has to do with the constraints that the hash of the "borrow" value and the hash of the original have to be equal. But it's also possible (in theory, at least) for a single type `T` to support multiple `Borrow<U>` implementations (i.e., you may have some `FooString` that implements `Borrow<str>` but also `Borrow<Vec<_>>` and so forth), so using `Borrow` may add flexibility.

I wonder if anyone on @rust-lang/libs might have thoughts here -- the question is, if we want to implement parallel operations on anything that can "yield up" a `str`, which of the various traits ought we to use? @cuviper selected `Borrow`, but `Deref` seems potentially suitable; I think @cuviper rejected `AsRef` because of the ambiguity it can induce:

> I also considered AsRef and AsMut, which are very similar to borrows, but I decided against it for the single fact that strings implement both AsRef<str> and AsRef<[u8]>. This would make it ambiguous if ParallelString and ParallelSlice shared any method names, and to drive that point home I went ahead and added par_split and par_split_mut on slices.

That same point might be used as an argument in favor of `Deref<Target = str>`, no?